kt/kt.go

```diff
@@ -315,6 +302,41 @@ func (c *Conn) doRPC(path string, values []KV) (code int, vals []KV, err error)
 	return resp.StatusCode, m, nil
 }
 
+func (c *Conn) roundTrip(method string, url *url.URL, headers http.Header, body []byte) (*http.Response, *time.Timer, error) {
+	req, t := c.makeRequest(method, url, headers, body)
+	resp, err := c.transport.RoundTrip(req)
+	if err == io.EOF {
+		// The connection was closed by the server while it was idle,
+		// which happens during graceful shutdown. Close all idle connections
+		// and retry once.
+		t.Stop()
+		c.transport.CloseIdleConnections()
+		req, t = c.makeRequest(method, url, headers, body)
+		resp, err = c.transport.RoundTrip(req)
+	}
+	if err != nil {
+		if !t.Stop() {
+			err = ErrTimeout
+		}
+		return nil, nil, err
+	}
+	return resp, t, nil
+}
+
+func (c *Conn) makeRequest(method string, url *url.URL, headers http.Header, body []byte) (*http.Request, *time.Timer) {
+	req := &http.Request{
+		Method:        method,
+		URL:           url,
+		Header:        headers,
+		Body:          ioutil.NopCloser(bytes.NewBuffer(body)),
+		ContentLength: int64(len(body)),
+	}
+	t := time.AfterFunc(c.timeout, func() {
+		c.transport.CancelRequest(req)
```

Comment:
This doesn't work with H/2, you could replace it with a much cleaner `Request.WithContext(context.WithTimeout)`